galsim/chromatic.py

```diff
@@ -1226,6 +1226,60 @@ def __init__(self, original, waves, oversample_fac=1.0, use_exact_sed=True,
 
         # Don't interpolate an interpolation.  Go back to the original.
         self.deinterpolated = original.deinterpolated
+    
+    @classmethod
+    def from_images(cls, images, waves, _force_stepk = None, _force_maxk = None, **kwargs):
+        """
+        Alternative initialization for InterpolatedChromaticObject from input images at specific wavelengths. Any parameters accepted by
+        the InterpolatedImage class can be passed in as kwargs. Note that stepk and maxk are parameters that can depend on the image size,
+        and therefore on the wavelength. If not given, as a list for every image, or a single number for all images, it will be calculated
+        using the input image pixel scale and dimensions. This means it will be identical for all images. This will cause small differences
+        from the normal use of this class using chromatic objects whose stepk and maxk are wavelength-dependent. 
+        
+        Parameters:
+            images:          list of Galsim Image objects to use as interpolated images.
+            wave:            list of wavelength values in nanometers.
+            _force_stepk:    list of step_k values to pass to InterpolatedImages for each image. Can also be single value
+                             to pass for all images alike. If not given, stepk is calculated from image pixel scale and dimensions. [Default: None]
+            _force_maxk:     list of max_k values to pass to InterpolatedImages for each image. Can also be single value
+                             to pass for all images alike. If not given, maxk is calculated from image pixel scale. [Default: None]
+
+        Returns:
+            An InterpolatedChromaticObject intialized from input images.
+        """
+        obj = cls.__new__(cls)  # Does not call __init__
+        obj.waves = np.array(waves) # images are assumed to be sorted by wavelength
+
+        # use_exact_sed set to false as input images won't have sed available. Ovsersample factor not relevant here, set to 1.0
+        obj.oversample = 1.0
+        obj.use_exact_sed = False
+        obj.separable = False
+        obj.interpolated = True
+        
+        # image properties
+        pix_scale = images[0].scale 
+        N = np.max(images[0].array.shape)
+        n_img = len(images)
+        
+        # default stepk and maxk from pixel scale
+        stepk  = np.full(n_img, 2*np.pi/(pix_scale*N))
+        maxk = np.full(n_img, np.pi/pix_scale)
```

Comment:
These aren't right I think.  Or at least, they are much simpler (so both less conservative in some respects but also less efficient in others) than what InterpolatedImage does to compute these.  See the functions `InterpolatedImage._getStepK` and `_getMaxK` for the more sophisticated calculation of these.  Most importantly, this stepk will usually be much smaller than required, and probably maxk will be larger than required.  This will make subsequent draw operations much slower than they need to be.

Some of the calculation of this can be a little slow, since it needs to do some calculations based on the actual profile in the image to get the right values.  But probably the resulting values for stepk and maxk are going to be nearly the same for all the images in something like this, so we can try to just do it once and use the same values for all the images here.

So I think you could make a regular InterpolatedImage from either the first image in the list, or maybe the sum of all of them (to get an average profile), let it compute stepk and maxk in its way with the sophisticated calculation, and then use those values of stepk and maxk for all of the objects here using the `_force` parameters.  That would probably work well and save a bunch of nearly equivalent calculations.

Reply:
Thanks for the explanation Mike. We initially chose to make this the default stepk, maxk because we had noticed some potential issues when using the stepk, maxk the `InterpolatedImage` calculates. The problem we were seeing was that when drawing images at a different pixel scale input images given to the `from_images` function, the image was being truncated after a certain pixel length, causing to truncate some features of the PSF (such as the spikes). As an example using the `roman.getPSF`, I did quick test to show this potential problem and would love to hear your opinions on this. I essentially test the `from_images` method using the stepk and maxk obtained from creating an average profile of the image and letting `InterpolatedImage` to calculate the stepk and maxk to use for all other input images. I then test the "conservative" stepk, maxk from the code snippet of this comment, and compare both from using the `roman.getPSF` directly. The image below shows the differences and the cut I had mentioned before. Let me know what you think of this, if needed I can also provide the exact code I used to generate these images, I haven't pushed the changes made to include the stepk, maxk from the average profile yet as I am still testing it here.
![from_images_problem](https://github.com/GalSim-developers/GalSim/assets/107272465/100eb22a-ed12-4059-95f1-28ac012257fc)

Reply:
Maybe to explain in a bit more detail what went into making the image above. I started from getting the PSF for the J band in Roman and then generating individual oversampled images at specfic wavelengths:
`scale = roman.pixel_scale/4
filt = 'Y106'
bandpass = roman_filters[filt]
# get roman PSF
psf_chrom = roman.getPSF(1, filt, n_waves = 10,  pupil_bin=8)
waves = psf_chrom.waves
psf_images = []
# get oversampled images at set wavelengths
for wave in waves:
    psf_wave = roman.getPSF(1, filt, wavelength=wave,  pupil_bin=8)
    #psf_chrom = roman.getPSF(1, filt, n_waves = 5,  pupil_bin=8)
    psf_img = psf_wave.drawImage(nx=480, ny=480, scale=scale, method = 'no_pixel')
    psf_images.append(psf_img)`

Then I create the `InterpolatedChromaticObject` using the two different ways of calculating stepk, maxk I described above. Note that in this version of the code I am using locally, I changed the default to be the average profile method proposed by Mike. 
`N = 480
# default stepk and maxk from pixel scale
stepk  = 2*np.pi/(scale*N)
maxk = np.pi/scale
    
# create flat SED and make interpolated object using default or conservative stepk, maxk
flat_sed = galsim.SED('1',wave_type = 'nm', flux_type = 'flambda' )
int_obj = galsim.InterpolatedChromaticObject.from_images(psf_images,  waves)
int_obj_conservative = galsim.InterpolatedChromaticObject.from_images(psf_images,  waves,
                                                                      _force_stepk =stepk,_force_maxk =maxk )
conv_obj = galsim.Convolve(int_obj, galsim.DeltaFunction()*flat_sed)
conv_obj_conservative = galsim.Convolve(int_obj_conservative, galsim.DeltaFunction()*flat_sed)`

Finally, I draw the all images at a different pixel scale from the one provided to the images:
`# draw images using 
drawing_scale = roman.pixel_scale
test_img = conv_obj.drawImage(bandpass, nx=100, ny=100, scale=drawing_scale, method = 'auto').array
test_img /= np.sum(test_img)

test_img_conservative = conv_obj_conservative.drawImage(bandpass, nx=100, ny=100, scale=drawing_scale, method = 'auto').array
test_img_conservative /= np.sum(test_img_conservative)

# true PSF using getPSF() object directly
true_obj = galsim.Convolve(psf_chrom, galsim.DeltaFunction()*flat_sed)
true_img = true_obj.drawImage(bandpass, nx=100, ny=100, scale=drawing_scale, method = 'auto').array
true_img /= np.sum(true_img)
`

Reply:
The knob we have in GalSim to control the level of approximation used for stepk is folding_threshold.  Smaller values mean large area in realspace.  

I think what you are effectively doing in your "conservative" version is the equivalent of lowering folding_threshold significantly.  If you try making the InterpolatedImage with `gsparams=galsim.GSParams(folding_threshold=1.e-3)`, I suspect you will get similar results.  

The advantage of conforming to this paradigm is that users can then be consistent in terms of the level of folding in k-space for the PSF and galaxy profiles when you convolve them together, rather than have one of them much more precise than the other, which would be wasted precision.  If you set folding_threshold for a Convolve, then it gets applied to all constituent profiles.

See also the docs about folding_threshold for the Roman PSF [here](http://galsim-developers.github.io/GalSim/_build/html/roman.html#galsim.roman.getPSF), and also [this PR](https://github.com/GalSim-developers/GalSim/pull/1089) where I show examples of the PSF with different folding_threshold and pupil_bin values.

Reply:
I don't think the reasons for using it here are different from other places in GalSim.  It's only going to be a cosmetic difference for very bright things.  When used with typical galaxy profiles, the default folding_threshold will be just fine for the vast majority of use cases.  E.g. in the Rubin-Roman sims, Troxel had a special setup for the bright stars so they looked pretty.  We have something similar in imsim.  But when used for galaxies, the regular gsparams are fine.

Reply:
Got it, I will try to push the changes discussed today.